drivers/javascript/cursor.coffee

```diff
@@ -211,50 +209,65 @@ class IterableResult
                 ).nodeify(cb)
         return @_closeCbPromise
 
-
-    _each: varar(1, 2, (cb, onFinished) ->
+    each: varar(1, 2, (cb, onFinished) ->
         unless typeof cb is 'function'
             throw new error.ReqlDriverError "First argument to each must be a function."
         if onFinished? and typeof onFinished isnt 'function'
             throw new error.ReqlDriverError "Optional second argument to each must be a function."
 
-        self = @
-        nextCb = (err, data) =>
-            if err?
-                if err.message is 'No more rows in the cursor.'
-                    onFinished?()
-                else
-                    cb(err)
-            else if cb(null, data) isnt false
-                @_next nextCb
-            else
-                onFinished?()
-        @_next nextCb
+        nextCb = =>
+            @_next().then (data) ->
+                return nextCb() if cb(null, data) isnt false
+            .catch (err) ->
+                return if err?.message is 'No more rows in the cursor.'
+
+        return nextCb().nodeify(onFinished)
     )
 
-    _eachAsync: varar(1, 2, (cb, errCb) ->
+    eachAsync: varar(1, 3, (cb, errCb, options = { concurrency: 1 }) ->
         unless typeof cb is 'function'
             throw new error.ReqlDriverError 'First argument to eachAsync must be a function.'
 
-        if errCb? and typeof errCb isnt 'function'
-            throw new error.ReqlDriverError "Optional second argument to eachAsync must be a function"
+        if errCb?
+            if typeof errCb is 'object'
+                options = errCb
+                errCb = undefined
+            else if typeof errCb isnt 'function'
+                throw new error.ReqlDriverError "Optional second argument to eachAsync must be a function or `options` object"
+
+        unless options and typeof options.concurrency is 'number' and options.concurrency > 0
+            throw new error.ReqlDriverError "Optional `options.concurrency` argument to eachAsync must be a positive number"
+
+        pending = []
+
+        userCb = (data) ->
+            return Promise.resolve(cb(data)) if cb.length <= 1 # either synchronous or awaits promise
+            return Promise.fromNode (handler) -> cb(data, handler) # callback-style async
 
         nextCb = =>
             if @_closeCbPromise?
                 return Promise.reject(new error.ReqlDriverError("Cursor is closed."))
             else
-                @_next().then (data) ->
-                    return cb(data) if cb.length <= 1 # either synchronous or awaits promise
-                    return Promise.fromNode (handler) -> cb(data, handler) # callback-style async
-                .then (result) ->
-                    return nextCb()
+                return @_next().then (data) ->
+                    return data if pending.length < options.concurrency
```

Comment:
`options.concurrency` will always either be what the user specified or `1` by the time we make it to this line.